### PR TITLE
Skia backend cleanup. Gpu support on Win32 via EGL.

### PR DESCRIPTION
**Pull request by MarchingCube**
**What does the pull request do?**
Idea for this PR was to support Skia gpu backends (OpenGL currently), this caused other changes in the backend that in the end resulted in quite massive refactoring. The goal is to have clean and maintainable render backend with architecture supporting gpu acceleration.

**What is the current behavior?**
There is no Skia gpu support, all the drawing is done through bitmaps. Also a lot of code parts were not easily readable, as some paths were more or less convoluted in their behavior.

**What is the updated/expected behavior with this PR?**
You can now enable gpu backend support on Windows. Most of the rendering will go through the gpu when enabled.

image

Due to refactoring I have also found and fixed some bugs:
  - TileBrush rendering with tiling set to none produces correct renders now
  - Line drawing with dashes actually renders dashes now
  - More Skia Render Tests are now active as more of them pass

**How was the solution implemented (if it's not obvious)?**

- Geometry
  - `GeometryImpl` now hosts the bulk of the geometry implementation
  - Hit testing now works as intended (fill and stroke)
  - Path intersection is now supported
- Media
  -  `BitmapImpl` was split into multiple parts:
      - `WriteableBitmap` - the only purely cpu writable bitmap (backed by `SkBitmap`)
      - `ImmutableBitmap` - read-only bitmap (backed by `SkImage`, can be gpu accelerated)

  - `FramebufferRenderTarget` is used in Raster backend, added surface caching
  - `WindowRenderTarget` - used in Gpu backend, usually backed by a window. Creates gpu accelerated `SkSurface` to handle drawing.
  - `SurfaceRenderTarget` - universal render target, can be either gpu or cpu based.
  - `DrawingContextImpl` is now the only drawing context, propagates gpu or cpu traits to layers it creates.
- Platform
  - Added `IEGLPlatform` and `IEGLContext` interface to main Avalonia platform. Idea for this is to provide EGL context to a user (I have seen some issues when people wanted to render via OpenGL to Avalonia window). Will need more work to support correctly, but I think this is step in this direction. Currently only Win32 has implementation for it (using custom EGL wrapper), as I am not that familiar with EGL on Linux or macOS. 
  - Skia has `IGpuRenderBackend` and `IGpuRenderContext`, Skia by itself is not aware of real rendering backend and I want to keep it this way, so this basically hides platform specifics from it.
- Renderers
  - I found (the hard way) that Avalonia is not cleaning up render targets when renderer is disposed. That means eventually GC will kick in and finalize Skia resources which leads to a crash (as finalizer thread does not have a OpenGL context current). To fix that I have added extra render target cleanup in both renderers.
- Misc
  - Updated SkiaSharp version to latest stable (1.60.0)
  - Added XML doc comments to almost everything (didn't touch `FormattedTextImpl` yet as it has some incoming PRs). 

**Checklist:**

- [x] Added unit tests (if possible)?
- [x] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Avaloniaui.net with user documentation

I have tested this with both Core2D and AvalonStudio and I have not noticed any regression. 
I still need to setup extra render tests project that will use gpu acceleration, as now all render tests are running against raster backend.
I would like people to try this PR with their own projects, and report if anything is wrong.

Related to #985

cc: @kekekeks I see that SkiaSharp.props is referencing `Avalonia.Skia.Linux.Natives` package from you that has no repo, or info how to distribute it.

### CODE REVIEW

**Comment by kekekeks** [comment]
>I see that SkiaSharp.props is referencing Avalonia.Skia.Linux.Natives package from you that has no repo, or info how to distribute it.

The package is built manually, because of the number of tricks that are required for a portable Linux binary. I will update it to 1.60.0 soon(-ish) since there also were several requests for it.


I haven't read the entire PR yet, but I have some comments regarding IOpenGlContext and IOpenGLPlatform.


You need to pass the list of "platform surfaces" instead of just `IPlatformHandle`. `HWND` is not a uniform concept.  For example, on Linux the source of the OpenGL context and the ways of obtaining it heavily depend on the actual windowing platform and (in case of GTK) on the underlying backend. In case of X11 we should be using GLX, in case of Wayland we should be using wayland-specific eglCreateWindowSurface call, in case of DRI/KMS/DRM we need some magic to actually get said "window". 

Another problem is that there could be different OpenGL implementations installed in the system. Even on Windows there is a "native" OpenGL support provided by the OS via vendor drivers and ANGLE that translates OpenGL calls to DirectX. The problem with the native one is that we basically can *not rely* on it, because it might be not available because of the lack of vendor-provided drivers installed in the system (ones from Windows Update usually don't have OpenGL support). Another issue with "native" OpenGL is interopability. We can't properly interop with DirectX code, which is required for GPU-accelerated rendering to non-windows surfaces backed by DirectX (e. g. our WPF compatibility layer that uses System.Windows.Interop.D3DImage for rendering). 
So we need to keep OpenGL implementation selection logic on windowing backend, which should provide a pointer to `glGetProcAddress`. SkiaSharp has support for constructing GrGlInterface from said pointer, I've specifically asked to add that a year ago.

It gets even more weird on Android/OSX/iOS, because there are restrictions on the *thread* that can make OpenGL calls to the render target. For efficient rendering we might need to render to a FBO first and transfer that FBO to the proper thread later. This logic might get tightly coupled with `GLKViewController` in case of iOS.

**Comment by kekekeks** [comment]
You are also managing OpenGL contexts in a wrong way. In case of EGL *everything* is using the same context, render target determined by the second argument of eglSwapBuffers. Also, the way of creating ResourceRender context is not portable. Some windowing platforms don't even have a concept of a window. `CreateWindow` just throws an exception.


See our [old GPU rendering implementation in C++](https://github.com/AvaloniaUI/libperspesk/blob/master/src/gl.cpp). It's a complete mess, but it takes into account the platform differences and has worked across Windows/Linux/iOS/Android.

**Comment by MarchingCube** [comment]
Thanks for the feedback. I am aware that current implementation is not portable, hence the WIP prefix. I will take a look at your c++ code and adjust the new implementation. 

For OpenGL context creation I completely agree that this is up to the windowing platform, and even currently I access it through an interface. But ultimately I would also like to give the user some way to control it, so we allow for customization in advanced scenario.

**Comment by MarchingCube** [comment]
If only GPU part is "controversial" I could cut it out from this PR and reintroduce it in another PR. As the gpu support is optional, and raster backend works just fine without it.

**Comment by kekekeks** [comment]
Unfortunately, we can't merge it until `Avalonia.Skia.Linux.Natives` package is updated, since it needs to be built for a particular version of SkiaSharp. I'll do that this week.

It might be worth to merge GPU support as is and modify it as we add support for other platforms. I also still need to thoroughly read the rest of the PR, again, hopefully somewhere this week.

**Comment by MarchingCube** [comment]
I have started working on EGL based context creation that should be much more portable. Should have something ready this week.

**Comment by MarchingCube** [comment]
Another update. I am already done with basic EGL implementation (still Win32, but more open right now). I have discussed with @kekekeks about Windows, and it seems we want to bundle ANGLE (https://github.com/google/angle) binaries to make sure everything works smoothly on all platforms. I already pinged @kekekeks about details, but I received no response yet. So in case somebody else can chime in and give some suggestion:
* we have x64 and x86 binaries for windows (can be built using Clang or MSVC)
* Avalonia.Win32 is supposed to provide those to the user

I read some info about nuget runtimes, etc, but I would like to get more details info so I don't waste too much time trying to figure that out.

**Comment by MarchingCube** [comment]
Updated the branch with new EGL based implementation.

**Comment by MarchingCube** [comment]
Progress update: Managed to iron out most of the bugs thanks to @danwalmsley testing. There is still some minor issue with resizing (when D3D11 is used as EGL backend) that I will try to track down.

**Comment by MarchingCube** [comment]
Seems the resizing issue happens on D2D backend as well. Due to window resize happening on other thread than actual rendering the sizes get out of sync and flickering/black borders happen. IMO fix for that should be done as a another PR.

**Review by jkoritzinsky — DISMISSED** [review]
I have one request for a change in here before we merge. Since everything that was added to `Avalonia.Base` has to do with rendering, can we move that code to `Avalonia.Visuals` since the rest of our rendering code is located there?

**Comment by MarchingCube** [comment]
@jkoritzinsky Moved the classes from `Avalonia.Base` to `Avalonia.Visuals` as requested. Right now I am also using natives package uploaded by @danwalmsley (https://www.nuget.org/packages/AvaloniaUI.Win32.Natives/), this could be replaced by official package (if it gets uploaded at some point).

Also summoning @kekekeks as I am more or less done with this PR.

File: src/Avalonia.Visuals/Rendering/DeferredRenderer.cs
Comment (grokys) [inline src/Avalonia.Visuals/Rendering/DeferredRenderer.cs]: Would be good to get the `DeferredRenderer` fixes in a separate PR too!

File: src/Skia/Avalonia.Skia/GeometryImpl.cs
Comment (grokys) [inline src/Skia/Avalonia.Skia/GeometryImpl.cs]: typo should be "changing"?

**Review by grokys — COMMENTED** [review]
Just a few suggestions, and a question.

Really appreciate the cleaning up of the SKia backend (we could also do with a cleanup of the Win32/Direct2D backends) but again, it would be best for @kekekeks to review these, hopefully he'll get time soon.

File: src/Avalonia.Visuals/Gpu/EGL.cs
Comment (nc4rrillo) [reply src/Avalonia.Visuals/Gpu/EGL.cs]: EGL is the glue behind the underlying platform Windowing system and the OpenGL API. It handles context and surface creation. While it is platform independent, it would tie Avalonia to at least knowing about GL. IMO it should probably live in an Avalonia.OpenGL project that backends can (optionally) reference.

File: src/Skia/Avalonia.Skia/Helpers/LambdaDisposable.cs
Comment (nc4rrillo) [inline src/Skia/Avalonia.Skia/Helpers/LambdaDisposable.cs]: Do we need this? There's a `Disposable` class in `System.Reactive` that we use in other places.

File: src/Skia/Avalonia.Skia/FormattedTextImpl.cs
Comment (nc4rrillo) [inline src/Skia/Avalonia.Skia/FormattedTextImpl.cs]: Was this something we were doing incorrectly before, or did something in the semantics of `BreakText` change between Skia versions?

**Pull request closed without merging**
